components/sup/src/manager/service/mod.rs

```diff
@@ -407,29 +412,46 @@ impl Service {
         self.supervisor.state == ProcessState::Down
     }
 
-    /// Compares the current state of the service to the current state of the census ring and
-    /// re-renders all templatable content to disk.
+    /// Compares the current state of the service to the current state of the census ring and the
+    /// user-config, and re-renders all templatable content to disk.
     ///
-    /// Returns true if any modifications were made.
+    /// Returns `true` if any modifications were made.
     fn update_templates(&mut self, census_ring: &CensusRing) -> bool {
         let census_group = census_ring.census_group_for(&self.service_group).expect(
             "Service update failed; unable to find own service group",
         );
-        let cfg_updated = self.cfg.update(census_group);
-        if cfg_updated || census_ring.changed() {
-            let (reload, reconfigure) = {
-                let ctx = self.render_context(census_ring);
-                let reload = self.compile_hooks(&ctx);
-                let reconfigure = self.compile_configuration(&ctx);
-                (reload, reconfigure)
-            };
-            self.needs_reload = reload;
-            self.needs_reconfiguration = reconfigure;
+        let cfg_updated_from_rumors = self.cfg.update(census_group);
+
+        if self.user_config_updated {
+            if let Err(e) = self.cfg.reload_user() {
+                outputln!(preamble self.service_group, "Loading user-config failed: {}", e);
+            }
         }
-        cfg_updated
+
+        let (cfg_changed, reload, reconfigure) = {
+            let ctx = self.render_context(census_ring);
+
+            if cfg_updated_from_rumors || census_ring.changed() {
+                let config_changed = cfg_updated_from_rumors || self.user_config_updated;
+                let needs_reload = self.compile_hooks(&ctx) || self.user_config_updated;
```

Comment:
I think compiling the hooks is all you need here... if a change to a user config file resulted in the compiled output of the templates, then that should be the only thing you care about, right?

Reply:
Couldn't a reload be necessary if e.g. a port was changed in the configuration of a DB?

Reply:
That should be taken care of when we generate a render context and compile the hooks. If the change to the user config resulted in different configs being rendered to disk, then that would be accounted for with the earlier code.

Reply:
Do you mean that the `reconfigure` hook would be run whenever the configuration in a `user.toml` has changed?
What if that hook isn't defined?

In general, I'm not sure I understand the connection between a change in the `user.toml` and hooks.
Couldn't there be an update to a `user.toml` file that doesn't produce any change in hooks?

Reply:
`user.toml` is one layer of a service's configuration (https://github.com/habitat-sh/habitat/blob/e2a946fafab06f7f4cfe8b5b4e883b88d69f356b/components/sup/src/manager/service/config.rs#L79-L80)

This, in turn, is part of the `RenderContext` (https://github.com/habitat-sh/habitat/blob/e2a946fafab06f7f4cfe8b5b4e883b88d69f356b/components/sup/src/templating/context.rs#L56-L63) that is created for templating both hooks and configuration.

When [hooks](https://github.com/habitat-sh/habitat/blob/e2a946fafab06f7f4cfe8b5b4e883b88d69f356b/components/sup/src/manager/service/mod.rs#L617) and [configuration](https://github.com/habitat-sh/habitat/blob/e2a946fafab06f7f4cfe8b5b4e883b88d69f356b/components/sup/src/manager/service/mod.rs#L598) are rendered according to the new data present in the `RenderContext`, we return a boolean that reflects whether any _content_ of the templated files changed. If they do, then we reconfigure or reload as needed.

If data changed, but the output in the templated files remained the same, then there is no reason to restart. In the extreme case, imagine adding comments to your `user.toml` file. The toml file certainly changed, but it will have no effect on the hooks or configuration of the service, by definition.